src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs

```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
-            }
+            // extract uncompressedSize from OP_COMPRESSED header in output
+            const int uncompressedSizeOffset = -(UncompressedSizeSize + CompressorTypeSize);
+            var position = output.Position;
+            output.Seek(uncompressedSizeOffset, SeekOrigin.Current); // rewind head
+            var uncompressedSize = ReadInt32(output);
+            output.Position = position; // return head to original position
+
+            var uncompressedMessage = new byte[uncompressedSize]; // does not include uncompressed message headers
+            input.ReadBytes(uncompressedMessage, offset: 0, count: uncompressedSize, CancellationToken.None);
+            var maxSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
+            var compressedBytes = new byte[maxSize];
+            var compressedSize = SnappyCodec.Compress(uncompressedMessage, 0, uncompressedSize, compressedBytes, 0);
+            output.Write(compressedBytes, 0, compressedSize);
 #else
             throw new NotSupportedException();
 #endif
         }
 
+        /// <summary>
+        /// Decompresses the message in <paramref name="input"/>, writing the uncompressed message to
+        /// <paramref name="output"/>.
+        /// </summary>
+        /// <param name="input">The input stream. Stream MUST HAVE read the header exactly up to and including
+        /// the compressorType aka CompressorId</param>
+        /// <param name="output">The output stream. Stream MUST HAVE written the message exactly up to and including
+        /// originalOpCode.</param>
+        /// <remarks>
+        /// The precondition for <paramref name="input"/> mirrors the precondition for <see cref="Compress"/>'s output
+        /// precondition.
+        /// The precondition for <paramref name="output"/> mirrors the precondition for <see cref="Compress"/>'s input
+        /// precondition.
+        /// </remarks>
         public void Decompress(Stream input, Stream output)
         {
 #if NET452 || NETSTANDARD2_0
-            using (var snappyStream = new SnappyStream(input, CompressionMode.Decompress, true))
-            {
-                snappyStream.CopyTo(output);
-            }
+            var position = input.Position;
+            input.Seek(0, SeekOrigin.Begin); // rewind head
+            const int headerSize =
+                MessageLengthSize + RequestIdSize + ResponseToSize + OpCodeSize +
+                OriginalOpCodeSize + UncompressedSizeSize + CompressorTypeSize;
+            var messageLength = ReadInt32(input) - headerSize;
+            input.Position = position; // return head to original position
+            var compressedMessage = new byte[messageLength];
+            input.ReadBytes(compressedMessage, offset: 0, count: messageLength, CancellationToken.None);
+            var decompressedMessage = SnappyCodec.Uncompress(compressedMessage);
+            output.Write(decompressedMessage, offset: 0, count: decompressedMessage.Length);
```

Comment:
I don't like the idea of *reading* from `output`.

Will this work instead:

            public void Decompress(Stream input, Stream output)
            {
    #if NET452 || NETSTANDARD2_0
                var compessedSize = (int)(input.Length - input.Position);
                var compressedBytes = new byte[compessedSize];
                input.ReadBytes(compressedBytes, offset: 0, count: compessedSize, CancellationToken.None);
                var decompressedBytes = SnappyCodec.Uncompress(compressedBytes);
                output.Write(decompressedBytes, offset: 0, count: decompressedBytes.Length);
    #else
                throw new NotSupportedException();
    #endif
            }

Reply:
Your suggestion is definitely cleaner and works. Fixed!